src/Compilers/CSharp/Portable/Symbols/TypeUnification.cs

```diff
@@ -78,14 +76,14 @@ private static bool CanUnifyHelper(TypeSymbol t1, TypeSymbol t2, ref MutableType
         /// Derived from Dev10's BSYMMGR::UnifyTypes.
         /// Two types will not unify if they have different custom modifiers.
         /// </remarks>
-        private static bool CanUnifyHelper(TypeWithAnnotations t1, TypeWithAnnotations t2, ref MutableTypeMap substitution)
+        private static bool CanUnifyHelper(TypeWithAnnotations t1, TypeWithAnnotations t2, ref MutableTypeMap? substitution)
         {
             if (!t1.HasType || !t2.HasType)
             {
                 return t1.IsSameAs(t2);
             }
 
-            if (TypeSymbol.Equals(t1.Type, t2.Type, TypeCompareKind.ConsiderEverything2) && t1.CustomModifiers.SequenceEqual(t2.CustomModifiers))
+            if (AreTypesAndCustomModifiersEqual(t1, t2))
```

Comment:
nit: we could move this check after the `if (substitution != null)`, so we only have that block once